src/toxcore/dht/daemon_state.rs

```diff
@@ -15,25 +13,96 @@ use toxcore::state_format::old::*;
 use toxcore::binary_io::*;
 use toxcore::dht::kbucket::*;
 
-/// Error that can happen when calling `deserialize_old` of DhtState.
-#[derive(Debug, Fail, PartialEq)]
-pub enum DeserializeError {
-    /// Error indicates that DhtState object can't be parsed.
-    #[fail(display = "Deserialize DhtState error: {:?}, packet: {:?}", error, data)]
+use std::fmt;
+
+use failure::{Backtrace, Context, Fail};
+use nom::{Needed, ErrorKind as NomErrorKind};
+
+/// An error that can occur while serializing/deserializing object
+#[derive(Debug)]
+pub struct DeserializeError{
+    ctx: Context<DeserializeErrorKind>,
+}
+
+impl DeserializeError{
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &DeserializeErrorKind {
+        self.ctx.get_context()
+    }
+
+    pub(crate) fn incomplete(needed: Needed, data: Vec<u8>) -> DeserializeError{
+        DeserializeError::from(DeserializeErrorKind::IncompleteData { needed, data })
+    }
+
+    pub(crate) fn deserialize(error: NomErrorKind, data: Vec<u8>) -> DeserializeError{
+        DeserializeError::from(DeserializeErrorKind::Deserialize { error, data })
+    }
+}
+
+impl Fail for DeserializeError{
+    fn cause(&self) -> Option<&Fail> {
+        self.ctx.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.ctx.backtrace()
+    }
+}
+
+impl fmt::Display for DeserializeError{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.ctx.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DeserializeErrorKind {
+    /// Error indicates that object can't be parsed.
     Deserialize {
         /// Parsing error
-        error: ErrorKind,
-        /// DhtState object serialized data
+        error: NomErrorKind,
+        /// object serialized data
         data: Vec<u8>,
     },
-    /// Error indicates that more data is needed to parse serialized DhtState object.
-    #[fail(display = "Bytes of DhtState object should not be incomplete: {:?}, data: {:?}", needed, data)]
+    /// Error indicates that more data is needed to parse serialized object.
     IncompleteData {
         /// Required data size to be parsed
         needed: Needed,
-        /// DhtState object serialized data
+        /// object serialized data
         data: Vec<u8>,
     },
+    /// This enum may grow additional variants, so this makes sure clients
+    /// don't count on exhaustive matching. (Otherwise, adding a new variant
+    /// could break existing code.)
+    #[doc(hidden)]
+    __Nonexhaustive,
```

Comment:
Where did you get it?

Reply:
I found it somewhere when I searched examples. But I don’t remember the exact url.